Setup project

XedinUnknown opened the pull request:
Just the initial project setup, so that contributors can get started easily, and for SoC.

- Composer;
- PHPUnit;
- Changelog and readme.
- Github Actions.
- This [builds][] on PHP 5.5 - 8.0.

[builds]: https://github.com/XedinUnknown/log-util/actions/workflows/continuous-integration.yml

XedinUnknown committed:
Setup project

- Composer;
- PHPUnit;
- Changelog and readme.
- Github Actions.
- Builds on PHP 5.5 - 8.0.

XedinUnknown committed:
Remove debug output from readme

XedinUnknown committed:
Add test bootstrap

XedinUnknown committed:
Install lowest deps on PHP 5.5 only

Jean85 commented on .github/workflows/continuous-integration.yml:
```diff
@@ -0,0 +1,36 @@
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        php-versions: ['5.5', '5.6', '7.0', '7.1', '7.2', '7.3', '7.4', '8.0']
```
Please use multiline array here, to make Git diff more readable

XedinUnknown replied:
✔️ 0fc3519

Jean85 commented on .github/workflows/continuous-integration.yml:
```diff
@@ -0,0 +1,36 @@
+        php-version: ${{ matrix.php-versions }}
+
+    - name: Analysing source code
+      run: find ./src/ ./inc/ ./tests/ -type f -name '*.php' -print0 | xargs -0 -L 1 -P 4 -- php -l
```
What does this do?

XedinUnknown replied:
This simply confirms that the PHP syntax isn't broken anywhere. But also, I didn't pay attention when pasting, and will clean this up to only look where is applicable for this package 👍

XedinUnknown replied:
✔️ 979e652

Jean85 commented on CHANGELOG.md:
```diff
@@ -0,0 +1,8 @@
+and this project adheres to [Semantic Versioning](http://semver.org/).
+
+## [[*next-version*]] - YYYY-MM-DD
+Initial release.
```
Can you add a reference to the original package and the PR that split this up?

Jean85 replied:
> Can you add a reference to the original package and the PR that split this up?

This seems the only unaddressed change request.

XedinUnknown replied:
✔️ 2244ac3

Jean85 commented on composer.json:
```diff
@@ -0,0 +1,36 @@
+    "psr/log": "^1.1.4"
+  },
+  "require-dev": {
+    "phpunit/phpunit": "^4.0 | ^5.0 | ^6.0 | ^7.0 | ^8.0 | ^9.0",
```
I doubt this will work on such a wide range of versions. Maybe we should release a 1.x with the same code as the original `psr/log` code, and afterwards a newer release where we bump toward newer releases.

XedinUnknown replied:
AFAIK, this is the _only_ way that it will work, because you need PHPUnit 4 to support PHP 5.4 and 5.5 at least. And it absolutely does work, as evidenced by [CI][] and what I explain in https://github.com/php-fig/log/pull/75#issuecomment-855381201.

[CI]: https://github.com/XedinUnknown/log-util/actions/runs/931184349

Jean85 replied:
I know that PHPUnit 4 is needed. I'm just not sure there's an easy way to make it work with such a wide constraint.

Point is, once you've declared a wide constraint and tagged it, we can't take it back, so we will not tag it if we can't assure it works.

Actions are not starting as it is, we will merge this and do subsequent PRs to fix stuff later.

XedinUnknown replied:
I imagine, we will remove PHPUnit versions prior to 8 from the 2.0 (or 0.2, depending on how you want to take this) version of this package, because legacy compat will not be a problem. And in 1.x there's probably no need to remove them, as they are required for legacy PHP versions, as you agreed.

Jean85 commented on composer.lock:
```diff
@@ -0,0 +1,1425 @@
+{
```
Do we really need to commit this?

XedinUnknown replied:
The benefit of committing the lock file is not so much for e.g. CI, since I do `update` there anyway (since this is a lib that will be installed in a variety of different dep configurations); but it is that anyone else can know exactly what works for the maintainers of the package (including sharing this between maintainers themselves), even if some other configurations for some reason don't.

But again, if you think this is not a good idea, I will remove.

Jean85 commented on tests/bootstrap.php:
```diff
@@ -0,0 +1,8 @@
+<?php
+
+call_user_func_array(function ($rootDir) {
```
Can we skip this and go directly using the autoloader as the bootstrap script? We can give PSR-4 and Composer for granted here.

XedinUnknown replied:
The idea behind the IIFE here is more about isolating scope and the variables in it, and this is just safer. With regard to taking Composer for granted, that's for sure, but checking first makes code even more portable, since the autoload script may reside in a root package of which this is a dependency. Because it's so easy to make this bit safer, I just do it everywhere, even if there's a very low risk factor.

So, if you confirm, of course I will remove this 😃

XedinUnknown committed:
Correct paths for PHP linting

XedinUnknown committed:
Use multiline list

XedinUnknown committed:
Build on PHP 5.3 and 5.4 as well

XedinUnknown committed:
Allow PHP 5.3

This is the minimal requirement from `psr/log`.

XedinUnknown committed:
Allow PHP 5.3 builds to fail

In order to support PHP 5.3, it would be necessary to downgrade to PHPCS 2.x, as 3.0-RC1 already bumps min PHP requirement to 5.4.
https://github.com/squizlabs/PHP_CodeSniffer/commit/161a3cfe6fb5c6cf324febd0f3c8395043cd35c2

However, PHPCS 2.x does not have support for PSR-12, which is the FIG standard, as it was only added in PHPCS 3.5
https://github.com/squizlabs/PHP_CodeSniffer/releases/tag/3.5.0

This change should ensure that we can at least formally build on PHP 5.3, even if it cannot pass, and documents this problem in history.

XedinUnknown committed:
Add reference to original decision

XedinUnknown committed:
Add missing flag

XedinUnknown committed:
Fix variable syntax

XedinUnknown committed:
Fix variable value syntax

XedinUnknown committed:
Revert "Fix variable value syntax"

This reverts commit dd24382a

XedinUnknown committed:
Add include for all extra variables used

CI was failing because `matrix.continue-on-error` was not being set.
https://github.com/XedinUnknown/log-util/actions/runs/939032673

XedinUnknown committed:
Remove potential var name ambiguity

XedinUnknown committed:
Declare PHP 5.3 builds as experimental

XedinUnknown commented:
@Jean85, I added PHP 5.3 and 5.4 to the matrix, so that this matches the spec. Unfortunatley, as I explain in 
2483bc1, it is not possible to make the build pass on PHP 5.3 right now. So I declared it as experimental, and it is allowed to fail.

Jean85 commented:
> @Jean85, I added PHP 5.3 and 5.4 to the matrix, so that this matches the spec. Unfortunatley, as I explain in
> [2483bc1](https://github.com/php-fig/log-util/commit/2483bc16b69f520747e7d85445c7429e2d119f18), it is not possible to make the build pass on PHP 5.3 right now. So I declared it as experimental, and it is allowed to fail.

You can overcome that limitation by removing the dependency during the build of that specific job, before calling `composer update`. I did something similar here: https://github.com/getsentry/sentry-symfony/blob/b9fbbdf6492e0571a39014b9ecd547093d697d86/.github/workflows/tests.yaml#L62-L63

XedinUnknown commented:
Ok, interesting workaround, thanks! Will try 👍

XedinUnknown committed:
Add PHPCS

XedinUnknown committed:
Downgrade PHPCS to 2.x on PHP 5.3

See 2483bc1.

XedinUnknown committed:
Downgrading corrected

- Now require-dev;
- Allow direct sub-dependency update.

XedinUnknown committed:
Downgrading corrected further

Now allowing update of all dependencies of the sub-dependency.

XedinUnknown committed:
Downgrading corrected even further

Now downgrading symfony/yaml also.

XedinUnknown committed:
Downgrading corrected even further

Now downgrading phpdocumentor/reflection-docblock also.

XedinUnknown committed:
Do not run PHPCS on PHP 5.3

XedinUnknown committed:
PHP 5.3 is no longer experimental

XedinUnknown committed:
Correct PHP 5.4 include

Was missing value for an option

XedinUnknown commented:
@Jean85, PHP 5.3 works now, but doesn't run PHPCS because PSR-12 does not exist there. But PHPCS on PHP 5.4 should be enough. Please take a look, and merge if all good 👍

XedinUnknown commented:
@Seldaek, it would be cool if there was a Composer flag like `-W` but that allows _any_ dependency to be up/downgraded, even if not a dependency of the package being `require`d. It may not be all that safe, but would sure save a lot of hassle :)

Jean85 commented on .github/workflows/continuous-integration.yml:
```diff
@@ -36,6 +40,10 @@ jobs:
       run: composer validate
 
+      # Because PHPCS 3 requires PHP 5.4
+    - run: composer require --dev -W squizlabs/php_codesniffer:^2.0 symfony/yaml:^2.0 phpdocumentor/reflection-docblock:^2.0
```
This seems an overkill to me: we don't need to run CS checks on all jobs, running it once should be more than enough.

We could run in in a separate job and not as part of the matrix, so we could get a dedicated check (and not having all of them fail when all tests are green and there's a space left somewhere).

XedinUnknown replied:
Thing is, syntax is different in different PHP versions, and PHPCS will check according to that syntax. So IMHO it makes sense to run PHPCS on as wide as possible range of PHP versions. But as usual, if you confirm, I will remove this.
